### PR TITLE
Workaround for container registry push/pull errors

### PR DESCRIPTION
**Pull request by KN4CK3R**
This PR addresses #19586

I added a mutex to the upload version creation which will prevent the push errors when two requests try to create these database entries. I'm not sure if this should be the final solution for this problem.

I added a workaround to allow a reupload of missing blobs. Normally a reupload is skipped because the database knows the blob is already present. The workaround checks if the blob exists on the file system. This should not be needed anymore with the above fix so I marked this code to be removed with Gitea v1.20.

### CODE REVIEW

**Comment by wxiaoguang** [comment]
I do not think it's right to use mutex. Database transaction should and could do that.

**Comment by wxiaoguang** [comment]
I proposed a PR:
* https://github.com/KN4CK3R/gitea/pull/3

And more thoughts: I do not think the workaround need to be remove in the future, keeping the workaround seems fine, and it will also help if the files are lost in some rare cases.

**Comment by ABastionOfSanity** [comment]
I put together a workaround based from 1.17.3 release.  This allows subsequent docker push to upload missing layers.

https://github.com/ABastionOfSanity/gitea/commit/00f3f3aedf9a629070bc9cacc4acc9a9d1f6e74b

**Comment by KN4CK3R** [comment]
> I do not think it's right to use mutex. Database transaction should and could do that (if #21836 could be written correctly)

I agree the mutex is not what we want but:

This is the error log of my code without the mutex:
```
...es/container/blob.go:47:func1() [E] [6377e7b6-18] Error inserting package: pq: duplicate key value violates unique constraint "UQE_package_s"
...es/container/blob.go:47:func1() [E] [6377e7b6-21] Error inserting package: pq: duplicate key value violates unique constraint "UQE_package_s"
...es/container/blob.go:47:func1() [E] [6377e7b6-20] Error inserting package: pq: duplicate key value violates unique constraint "UQE_package_s"
...ntainer/container.go:86:apiError() [E] [6377e7b6-18] pq: duplicate key value violates unique constraint "UQE_package_s"
...ntainer/container.go:86:apiError() [E] [6377e7b6-20] pq: duplicate key value violates unique constraint "UQE_package_s"
```
The `func1` error is the call to `packages_model.TryInsertPackage`.

And this the error log of your code without the mutex:
```
...es/container/blob.go:47:func1() [E] [6377e861-19] Error inserting package: pq: current transaction is aborted, commands ignored until end of transaction block
...es/container/blob.go:47:func1() [E] [6377e861-21] Error inserting package: pq: current transaction is aborted, commands ignored until end of transaction block
...es/container/blob.go:47:func1() [E] [6377e861-22] Error inserting package: pq: current transaction is aborted, commands ignored until end of transaction block
...es/container/blob.go:47:func1() [E] [6377e861-20] Error inserting package: pq: current transaction is aborted, commands ignored until end of transaction block
...ntainer/container.go:86:apiError() [E] [6377e861-19] pq: current transaction is aborted, commands ignored until end of transaction block
...ntainer/container.go:86:apiError() [E] [6377e861-21] pq: current transaction is aborted, commands ignored until end of transaction block
```
This error is returned by the call to `Get` inside the `Insert` error handler. So, different error but still does not work.

**Comment by wxiaoguang** [comment]
OK, I know the problem now. Will try to propose a fix later (today)

**Comment by wxiaoguang** [comment]
The problem of my PR is: Only PostgreSQL can not continue current transaction if an SQL error occurs (MySQL/MSSQL/SQLite3 can)

PgSQL can use savepoint to skip the failed SQL and continue the current transaction, but I do not know whether XORM have general SAVEPOINT support.

Without a general SAVEPOINT support, I am afraid there is no clear solution to this problem (that's why GitLab finally only supports one database).

Another approach is to write various SQL dialects for `INSERT IGNORE`.

And one more approach, as a workaround without mutex, if you agree, we can move the `TryInsert` out the transaction, just insert and ignore the error, and do `Get` later, then there will be always a record.

**Review by wxiaoguang — APPROVED** [review]
Although the mutex is not the proper solution, making the transaction work correctly on all databases need a lot of time.

So since this PR is a workaround, I will vote my LGTM for current approach if there are enough comments for improvement in the future.

File: routers/api/packages/container/blob.go [hunk 1]
```diff
@@ -28,6 +34,9 @@ func saveAsPackageBlob(hsr packages_module.HashedSizeReader, pi *packages_servic
 
 	contentStore := packages_module.NewContentStore()
 
+	var uploadVersion *packages_model.PackageVersion
+
+	uploadVersionMutex.Lock()
```
Comment (wxiaoguang) [inline routers/api/packages/container/blob.go:41]: As a workaround, I think here needs enough comments, and a `FIXME` tag.

**Comment by lunny** [comment]
Could you copy the tests code from my PR?

**Comment by KN4CK3R** [comment]
I don't see tests in #21836

**Comment by lunny** [comment]
> #21836

Sorry. Maybe we could have some parallel upload in https://github.com/go-gitea/gitea/blob/main/tests/integration/api_packages_container_test.go#L155-L157

**Comment by KN4CK3R** [comment]
Like that?
```go
var wg sync.WaitGroup
for i := 0; i < 10; i++ {
	wg.Add(1)
	go func() {
		defer wg.Done()
		upload(i)
	}()
}
wg.Wait()
```
Does that add a benefit? I think tests should be reproducible but this test could fail or not fail depending on whatever.

**Comment by wxiaoguang** [comment]
```
var testSlowTransaction bool

db.WithTx({
   insert
   if testSlowTransaction  {
       Sleep(1s)
   }
});
```

then 100% reproducible.

**Comment by lunny** [comment]
> Like that?
> 
> ```go
> var wg sync.WaitGroup
> for i := 0; i < 10; i++ {
> 	wg.Add(1)
> 	go func() {
> 		defer wg.Done()
> 		upload(i)
> 	}()
> }
> wg.Wait()
> ```
> 
> Does that add a benefit? I think tests should be reproducible but this test could fail or not fail depending on whatever.

I think yes. Why this test maybe failed? Maybe only for sqlite?

**Comment by KN4CK3R** [comment]
No, sqlite is the only database where this never fails. I just wanted to say that a succesfull test does not indicate correct code in this case. I added parallel uploads in the integration test and with Postgresql this fails instantly and works fine with the patch. Will push the change later.

**Comment by wxiaoguang** [comment]
How to reproduce:

transaction A (t-A): 1. insert a key-X 3. do other long-time work. 4. commit
transaction B (t-B): 2. insert the key-X, then it blocks until 4 finishes, then  error occurs.
timeline: 1 (t-A inserts), 2 (t-B inserts), 3 (t-A do long-time work, eg: uploading or sleeping), 4 (t-A commit), then  t-B's insert will return error.

So, the Sleep  will make reproduce the problem.  And it can use some variables to detect such problem occurs and is  handled properly.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
make L-G-T-M work

**Pull request merged**